### PR TITLE
Add --train_steps_to_log CLI option and deduplicate training/validation/test steps

### LINKED ISSUE #674: [Feature] Support per-step loss logging during training
Currently, during training over multiple autoregressive steps, the overall training loss is aggregated into a single scalar and logged as `train_loss`. There is no way to monitor the degradation of loss at individual unroll steps during training (unlike validation and testing, which support per-step loss logging using `val_loss_unroll{step}`).

This enhancement proposes adding a `--train_steps_to_log` CLI option (similar to `--val_steps_to_log`) to allow users to log the training loss for specific unroll steps separately (e.g. `train_loss_unroll1`, `train_loss_unroll5`, `train_loss_unroll10`).

### PR DESCRIPTION
**Pull request by GiGiKoneti**
## Describe your changes

- Add the optional `--train_steps_to_log` CLI parameter to enable selective logging of training loss at individual unroll steps (e.g., `train_loss_unroll1`, `train_loss_unroll5`), defaulting to `[]`.
- Encapsulate prediction, standard deviation fallback resolving, and step-wise loss computation logic into a shared helper `_compute_prediction_and_loss`.
- Refactor `training_step`, `validation_step`, and `test_step` to use the helper, eliminating duplicate code between the train and val loops.
- Add unit tests verifying config options forwarding and step limit validation.

## Issue Link

closes #674

## Type of change

- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by GiGiKoneti** [comment]
@joeloskarsson @sadamov 
ready for your review..kindly lemme know what all edits/corecctions would it make it better.

File: neural_lam/models/module.py [hunk 1]
```diff
@@ -234,6 +240,7 @@ def __init__(
         # Warn once per phase if val_steps_to_log exceeds the actual rollout
         self._val_steps_warn_issued = False
         self._test_steps_warn_issued = False
+        self._train_steps_warn_issued = False
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: I think we can make this a dict, to not have to make three separate variables.

File: neural_lam/models/module.py [hunk 2]
```diff
@@ -418,37 +429,65 @@ def all_gather_cat(self, tensor_to_gather):
...
+        if phase == "train":
+            steps_to_log = (
+                hparams.train_steps_to_log  # type: ignore[attr-defined]
+            )
+            arg_name = "--train_steps_to_log"
+        else:
+            steps_to_log = (
+                hparams.val_steps_to_log  # type: ignore[attr-defined]
+            )
+            arg_name = "--val_steps_to_log"
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: Should be able to just use `phase` to construct these for train/val/test, not needing an if-statement.

File: neural_lam/models/module.py [hunk 3]
```diff
@@ -418,37 +429,65 @@ def all_gather_cat(self, tensor_to_gather):
...
         setattr(self, flag, True)
 
-    def validation_step(self, batch, batch_idx):
+    def _compute_prediction_and_loss(
+        self,
+        batch: tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor],
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, int]:
         """
-        Perform a single validation step.
+        Compute prediction, targets, standard deviation, and step-wise loss.
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]:
```suggestion
        Compute predicted mean, standard deviation, and step-wise loss. Also extract and return corresponding target from batch.
```
Since target is not something computed.

File: neural_lam/models/module.py [hunk 4]
```diff
@@ -418,37 +429,65 @@ def all_gather_cat(self, tensor_to_gather):
...
-            The index of the batch.
+
+        Returns
+        -------
+        prediction : torch.Tensor
+            Model predictions.
+        target_states : torch.Tensor
+            Target states.
+        pred_std : torch.Tensor
+            Predicted/clamped standard deviation.
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: I don't understand "clamped"? It's either predicted or coming from the pre-defined `self.per_var_std` right?

File: neural_lam/models/module.py [hunk 5]
```diff
@@ -418,37 +429,65 @@ def all_gather_cat(self, tensor_to_gather):
...
+        time_step_loss : torch.Tensor
+            Loss for each unroll step.
+        batch_size : int
+            The batch size.
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: I feel like returning the batch size feels a little out of place here. Would rather keep that separate, either just directly in `train_step` and `validation_step`, or as a separate method.

File: neural_lam/models/module.py [hunk 6]
```diff
@@ -473,7 +538,7 @@ def validation_step(self, batch, batch_idx):
             on_step=False,
             on_epoch=True,
             sync_dist=True,
-            batch_size=batch[0].shape[0],
+            batch_size=batch_size,
         )
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: This is very similar to the code doing the same thing in the train_step. I think we need to refactor this more to avoid duplicate code here.,

File: neural_lam/models/module.py [hunk 7]
```diff
@@ -372,27 +379,31 @@ def training_step(self, batch):
...
+        batch_loss = torch.mean(time_step_loss)
+        self._warn_skipped_steps(len(time_step_loss), "train")
 
         log_dict = {"train_loss": batch_loss}
+        hparams = self.hparams
+        train_steps_to_log = (
+            hparams.train_steps_to_log  # type: ignore[attr-defined]
+        )
+        if train_steps_to_log:
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: Is this if-statement really needed? The default `None` is changed to an empty list and that would make the for-loop void.

File: neural_lam/models/module.py (hunk 6 above)
Comment (joeloskarsson) [reply neural_lam/models/module.py]: Similar code is also found in test step.

File: tests/test_train_model_warnings.py [hunk 8]
```diff
@@ -87,3 +95,33 @@ def capture_init(_self, **kwargs):
         "create_gif" in captured_kwargs
     ), "create_gif was not forwarded to ForecasterModule"
     assert captured_kwargs["create_gif"] is True
+    assert (
+        "train_steps_to_log" in captured_kwargs
+    ), "train_steps_to_log was not forwarded to ForecasterModule"
+    assert captured_kwargs["train_steps_to_log"] == [2]
+
+
+def test_train_steps_to_log_validation():
```
Comment (joeloskarsson) [inline tests/test_train_model_warnings.py]: I think we should handle testing for `--train_steps_to_log` and `--val_steps_to_log` together, e.g. as a parametrized test.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Thanks to looking at this! This leads to some more general improvement and de-duplication of code in the ForecasterModule. As it touches a number of things in there I had some comments about exactly how we should go about this.

File: neural_lam/models/module.py [hunk 9]
```diff
@@ -418,37 +412,59 @@ def all_gather_cat(self, tensor_to_gather):
...
+        if self._steps_warn_issued[phase]:
             return
-        invalid = [s for s in self.hparams.val_steps_to_log if s > pred_steps]
+        log_hparams = {
+            "train": "train_steps_to_log",
+            "val": "val_steps_to_log",
+            "test": "val_steps_to_log",
+        }
+        hparam_name = log_hparams[phase]
+        arg_name = f"--{hparam_name}"
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: This does not need to be defined as a variable I think.

File: neural_lam/models/module.py [hunk 10]
```diff
@@ -418,37 +412,59 @@ def all_gather_cat(self, tensor_to_gather):
...
+        """Warn once per phase if any log steps exceed pred_steps."""
+        if self._steps_warn_issued[phase]:
             return
-        invalid = [s for s in self.hparams.val_steps_to_log if s > pred_steps]
+        log_hparams = {
+            "train": "train_steps_to_log",
+            "val": "val_steps_to_log",
+            "test": "val_steps_to_log",
+        }
+        hparam_name = log_hparams[phase]
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]:
```suggestion
        hparam_name = "train_steps_to_log" if phase == "train" else "val_steps_to_log"
```

File: neural_lam/models/module.py [hunk 11]
```diff
@@ -418,37 +412,59 @@ def all_gather_cat(self, tensor_to_gather):
...
+        Returns
+        -------
+        prediction : torch.Tensor
+            Model predictions.
+        target_states : torch.Tensor
+            Target states.
+        pred_std : torch.Tensor
+            Predicted or pre-defined standard deviation.
+        time_step_loss : torch.Tensor
+            Loss for each unroll step.
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: Write out shapes of these tensors

File: neural_lam/models/module.py [hunk 12]
```diff
@@ -459,23 +475,72 @@ def validation_step(self, batch, batch_idx):
...
+        batch_idx : int
+            The index of the batch.
+        """
+        prediction, target_states, pred_std, time_step_loss = (
+            self._compute_prediction_and_loss(batch)
+        )
+        mean_loss = torch.mean(time_step_loss)
+        batch_size = batch[0].shape[0]
+
+        self._warn_skipped_steps(len(time_step_loss), "val")
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: Would it be possible to move this warning check into `_log_step_loss` ?

File: neural_lam/models/module.py [hunk 13]
```diff
@@ -459,23 +475,72 @@ def validation_step(self, batch, batch_idx):
...
+        phase: str,
+        batch_size: int,
+    ) -> None:
+        """Log mean and step-wise losses for a given phase."""
+        prefixes = {
+            "train": "train_loss",
+            "val": "val_mean_loss",
+            "test": "test_mean_loss",
+        }
+        log_dict = {prefixes[phase]: mean_loss}
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: This seems like it should be possible to simplify, so we don't have to define a dict. Many lines of code for a simple mapping between strings.

File: neural_lam/models/module.py [hunk 14]
```diff
@@ -459,23 +475,72 @@ def validation_step(self, batch, batch_idx):
...
+
+        log_hparams = {
+            "train": "train_steps_to_log",
+            "val": "val_steps_to_log",
+            "test": "val_steps_to_log",
         }
-        val_log_dict["val_mean_loss"] = mean_loss
+        hparam_name = log_hparams[phase]
```
Comment (joeloskarsson) [inline neural_lam/models/module.py]: Simplify, see my other similar suggestion.

File: neural_lam/train_model.py [hunk 15]
```diff
@@ -368,6 +375,14 @@ def main(input_args=None):
...
                 "--val_steps_to_log."
             )
+    # Check --train_steps_to_log
+    for step in args.train_steps_to_log:
+        if step > args.ar_steps_train:
+            raise ValueError(
+                f"Can not log training step {step} when training is "
+                f"only unrolled {args.ar_steps_train} steps. Adjust "
+                "--train_steps_to_log."
+            )
```
Comment (joeloskarsson) [inline neural_lam/train_model.py]: Seems like we could do this in a loop over `[(args.val_steps_to_log, args.ar_steps_eval),(args.train_steps_to_log, args.ar_steps_train)]` to avoid duplicate code.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Thanks for updating! Had another look over after the changes, with some suggestions.

File: neural_lam/train_model.py [hunk 16]
```diff
@@ -360,14 +367,17 @@ def main(input_args=None):
...
+        (args.val_steps_to_log, args.ar_steps_eval, "validation", "val"),
+        (args.train_steps_to_log, args.ar_steps_train, "training", "train"),
+    ]:
+        for step in steps:
+            if step > max_steps:
+                raise ValueError(
+                    f"Can not log {phase_name} step {step} when {phase_name} "
+                    f"is only unrolled {max_steps} steps. Adjust "
+                    f"--{flag_prefix}_steps_to_log."
+                )
```
Comment (joeloskarsson) [inline neural_lam/train_model.py]: Hmm, I know this follows my previous suggestion, but with the additional iteration vars this feels unnecessarily complicated, and gets hard to read. Something like this should work I think:
```suggestion
    for phase, max_steps in [
        ("train" args.ar_steps_train),
        ("val", args.ar_steps_eval),
    ]:
        for step in getattr(args, f"{phase}_steps_to_log"):
            if step > max_steps:
                raise ValueError(
                    f"Can not log {phase} step {step} when "
                    f"only unrolling {max_steps} steps during {phase} phase. Adjust "
                    f"--{phase}_steps_to_log."
                )
```
(I did not test this, so might require some adjusting. Hopefully it at least gets the idea across 😄).

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Good fixes, I think this looks great now. Only had one nitpick suggestion, then I think this is good to go!

**Comment by joeloskarsson** [comment]
Loos to also be missing some parts of the PR template (only noticed this now). Pls re-add those so i can check the required boxes there.

**Comment by GiGiKoneti** [comment]
@joeloskarsson 
I have update the PR body using full template. And did requested changes

**Review by joeloskarsson — APPROVED** [review]
Thanks, this looks good now! A small addition that brought with it a lot of refactoring in the module code, that I think made it cleaner and easier to maintain in the end!

This should go through the next dev meeting, but code is all good imo.
